GEOWAVE-1723: "filesystem" datastore with pluggable data formatter

rfecher opened the pull request:
See issue #1723 for details.

Signed-off-by: Rich Fecher <rich.fecher@blacklynx.tech>

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/FileSystemDataFormatter.java:
```diff
@@ -0,0 +1,312 @@
+package org.locationtech.geowave.datastore.filesystem;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/FileSystemDataFormatterRegistry.java:
```diff
@@ -0,0 +1,35 @@
+package org.locationtech.geowave.datastore.filesystem;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/FileSystemDataFormatterSpi.java:
```diff
@@ -0,0 +1,11 @@
+package org.locationtech.geowave.datastore.filesystem;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/FileSystemDefaultConfigProvider.java:
```diff
@@ -0,0 +1,30 @@
+   * Create the properties for the config-properties file
+   */
+  private void setProperties() {
+    configProperties.setProperty("store.default-rocksdb.opts.gwNamespace", "default");
```
There are `rocksdb` references in both of these properties.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/cli/FileSystemOperationProvider.java:
```diff
@@ -0,0 +1,14 @@
+package org.locationtech.geowave.datastore.filesystem.cli;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/cli/FileSystemSection.java:
```diff
@@ -0,0 +1,12 @@
+package org.locationtech.geowave.datastore.filesystem.cli;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/cli/ListFormatsCommand.java:
```diff
@@ -0,0 +1,36 @@
+package org.locationtech.geowave.datastore.filesystem.cli;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/config/FileSystemOptions.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Parameter(
+      names = "--format",
+      description = "Optionally use a formatter configured with Java SPI of type org.locationtech.geowave.datastore.filesystem.FileSystemDataFormatterSpi.  Defaults to \""
```
As with the comment in the documentation, it's probably best to leave implementation details out of the CLI help.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/operations/FileSystemOperations.java:
```diff
@@ -0,0 +1,260 @@
+                    : options.getGeoWaveNamespace())).toString();
+
+    visibilityEnabled = options.getStoreOptions().isVisibilityEnabled();
+    // a factory method that returns a RocksDB instance
```
RocksDB mention.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/util/FileSystemIndexKeyWrapper.java:
```diff
@@ -0,0 +1,31 @@
+package org.locationtech.geowave.datastore.filesystem.util;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/util/FileSystemKey.java:
```diff
@@ -0,0 +1,17 @@
+package org.locationtech.geowave.datastore.filesystem.util;
```
Missing copyright header.

jdgarrett commented on extensions/datastores/filesystem/src/main/java/org/locationtech/geowave/datastore/filesystem/util/FileSystemUtils.java:
```diff
@@ -0,0 +1,243 @@
+package org.locationtech.geowave.datastore.filesystem.util;
```
Missing copyright header.

jdgarrett commented on test/src/main/java/org/locationtech/geowave/test/FileSystemStoreTestEnvironment.java:
```diff
@@ -0,0 +1,65 @@
+  @Override
+  public void tearDown() throws Exception {
+    // this helps clean up any outstanding native resources
+    RocksDBClientCache.getInstance().closeAll();
```
RocksDB reference.

jdgarrett commented on test/src/test/java/org/locationtech/geowave/test/secondary/BasicSecondaryIndexIT.java:
```diff
@@ -36,7 +36,11 @@
+          GeoWaveStoreType.ROCKSDB,
+      // testLocalIngestAndQuerySpatialAndSpatialTemporal sporadically fails on
+      // testDeleteByBasicQuery
+      // GeoWaveStoreType.FILESYSTEM
```
Is this meant to be fixed at some point in the future?

jdgarrett commented on test/src/test/java/org/locationtech/geowave/test/secondary/MapReduceSecondaryIndexIT.java:
```diff
@@ -46,7 +46,11 @@
+          GeoWaveStoreType.ROCKSDB,
+      // testLocalIngestAndQuerySpatialAndSpatialTemporal sporadically fails on
+      // testDeleteByBasicQuery
+      // GeoWaveStoreType.FILESYSTEM
```
Is this meant to be fixed at some point in the future?

jdgarrett commented on test/src/test/java/org/locationtech/geowave/test/secondary/MapReduceSecondaryIndexIT.java:
```diff
@@ -62,7 +66,11 @@
+          GeoWaveStoreType.ROCKSDB,
+      // testLocalIngestAndQuerySpatialAndSpatialTemporal sporadically fails on
+      // testDeleteByBasicQuery
+      // GeoWaveStoreType.FILESYSTEM
```
Is this meant to be fixed at some point in the future?

rfecher replied:
it *never* happens for me locally but is consistently happening on travis for some reason.  I don't know, of course it would be nice to have everything pass (its just towards the end, the deletion part, on one of the tests, the one with both spatial and temporal indices)

jdgarrett commented on test/src/test/java/org/locationtech/geowave/test/services/GeoServerIngestIT.java:
```diff
@@ -87,7 +87,10 @@
-          GeoWaveStoreType.ROCKSDB},
+          GeoWaveStoreType.ROCKSDB,
+          // filesystem sporadically fails with a null response on spatial-temporal subsampling
+          // (after the spatial index is removed and the services restarted)
```
Will this happen enough to cause frequent build failures on CI?

rfecher replied:
It never happens for me locally but is consistently happening on travis for some reason, I'm assuming its related to the secondary index IT tests failing deletion with both indices, but seems strange to not ever happen locally.

jdgarrett commented on extensions/datastores/filesystem/pom.xml:
```diff
@@ -0,0 +1,45 @@
+		<version>1.2.0-SNAPSHOT</version>
+	</parent>
+	<artifactId>geowave-datastore-filesystem</artifactId>
+	<name>GeoWave RocksDB</name>
```
Wrong artifact name.

jdgarrett commented on services/rest/src/main/java/org/locationtech/geowave/service/rest/operations/AddFileSystemStoreCommand.java:
```diff
@@ -0,0 +1,138 @@
+import org.locationtech.geowave.core.store.cli.store.DataStorePluginOptions;
+import org.locationtech.geowave.core.store.cli.store.StoreSection;
+import org.locationtech.geowave.datastore.filesystem.config.FileSystemOptions;
+import org.locationtech.geowave.datastore.rocksdb.config.RocksDBOptions;
```
RocksDB import.

jdgarrett commented on test/src/main/java/org/locationtech/geowave/test/FileSystemStoreTestEnvironment.java:
```diff
@@ -0,0 +1,65 @@
+import org.locationtech.geowave.core.store.api.DataStore;
+import org.locationtech.geowave.datastore.filesystem.FileSystemStoreFactoryFamily;
+import org.locationtech.geowave.datastore.filesystem.config.FileSystemOptions;
+import org.locationtech.geowave.datastore.rocksdb.util.RocksDBClientCache;
```
RocksDB import.

rfecher commented on docs/content/commands/manpages/store/geowave-addstore.txt:
```diff
@@ -80,26 +80,37 @@ When the `redis` type option is used, additional options are:
+  The directory to read/write to.  Defaults to "geowave" in the working directory.
+
+*--format* _<format>_::
+  Optionally use a formatter configured with Java SPI of type org.locationtech.geowave.datastore.filesystem.FileSystemDataFormatterSpi.  Defaults to 'binary' which is a compact geowave serialization.  Use `geowave util filesystem listformats` to see available formats.
```
yeah, I get that and I can leave that out. But this is an odd one because the only "option" is "binary" without going the SPI route so it is pretty much a non-option without some explanation about how to get additional options through SPI.

rfecher commented:
I addressed the concerns.  See comments regarding ITs failing, but they always pass locally for me so it is isolated to travis and more challenging to diagnose.  I am running 2 passes of the rest of the tests on travis, the last one with secondary indexing enabled on all the tests so overall its pretty solid.  But yes, I'd really prefer all tests to pass all the time on travis-ci.  I can poke around more in the future if I have a spare cycle and renyone is welcome to poke around.

rfecher commented:
good catch - fixed those rocks references and squashed it back down to a single commit
